### PR TITLE
DWARF5 line header parsing

### PR DESCRIPTION
**Pull request by MewtR**
- [ ] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**

My attempt to support DWARF5 line header parsing. Mostly done, just need to update the tests.

### CODE REVIEW

**Comment by trufae** [comment]
Ready to review now? I didnt expect that pr and thats beautiful! 🥰

**Comment by MewtR** [comment]
Thanks, yeah I think now it's fine. Let me know what you think.

**Comment by trufae** [comment]
Fix those warnings
Screenshot 2023-01-16 at 00 43 13

File: libr/bin/dwarf.c [hunk 1]
```diff
@@ -554,34 +556,159 @@ static const ut8 *parse_line_header_source(RBinFile *bf, const ut8 *buf, const u
...
 // TODO DWARF 5 line header parsing, very different from ver. 4
 // Because this function needs ability to parse a lot of FORMS just like debug info
 // I'll complete this function after completing debug_info parsing and merging
 // for the meanwhile I am skipping the space.
 static const ut8 *parse_line_header_source_dwarf5(RBinFile *bf, const ut8 *buf, const ut8 *buf_end,
-	RBinDwarfLineHeader *hdr, Sdb *sdb, int mode) {
-// 	int i = 0;
-// 	size_t count;
-// 	const ut8 *tmp_buf = NULL;
+	RBinDwarfLineHeader *hdr, Sdb *sdb, int mode, PrintfCallback print) {
```
Comment (trufae) [inline libr/bin/dwarf.c:564]: function signatures must be always 1 line for grepping purposes

File: libr/bin/dwarf.c [hunk 2]
```diff
@@ -554,34 +556,159 @@ static const ut8 *parse_line_header_source(RBinFile *bf, const ut8 *buf, const u
...
+	size_t maxlen = 0xfff;
+	int i, j;
+	enum type { DIRECTORIES,
+		FILES };
 
-// 	ut8 dir_entry_count = READ8 (buf);
-// 	// uleb128 pairs
-// 	ut8 dir_count = READ8 (buf);
+	for (i = DIRECTORIES; i <= FILES; i++) {
+
```
Comment (trufae) [inline libr/bin/dwarf.c:571]:
```suggestion
```

File: libr/bin/dwarf.c [hunk 3]
```diff
@@ -554,34 +556,159 @@ static const ut8 *parse_line_header_source(RBinFile *bf, const ut8 *buf, const u
...
+		}
 
-// beach:
-// 	sdb_free (sdb);
+		buf = r_uleb128 (buf, buf_end - buf, &total_entries, NULL);
 
-	return NULL;
+		if (i == FILES) {
+			if (total_entries > 0) {
+				hdr->file_names = calloc (sizeof (file_entry), total_entries);
```
Comment (trufae) [inline libr/bin/dwarf.c:593]: check if calloc fails

File: libr/bin/dwarf.c [hunk 4]
```diff
@@ -554,34 +556,159 @@ static const ut8 *parse_line_header_source(RBinFile *bf, const ut8 *buf, const u
...
 // Because this function needs ability to parse a lot of FORMS just like debug info
 // I'll complete this function after completing debug_info parsing and merging
 // for the meanwhile I am skipping the space.
 static const ut8 *parse_line_header_source_dwarf5(RBinFile *bf, const ut8 *buf, const ut8 *buf_end,
-	RBinDwarfLineHeader *hdr, Sdb *sdb, int mode) {
-// 	int i = 0;
-// 	size_t count;
-// 	const ut8 *tmp_buf = NULL;
+	RBinDwarfLineHeader *hdr, Sdb *sdb, int mode, PrintfCallback print) {
+	size_t maxlen = 0xfff;
```
Comment (trufae) [inline libr/bin/dwarf.c:565]:
```suggestion
	const size_t maxlen = 0xfff;
```

File: libr/bin/dwarf.c [hunk 5]
```diff
@@ -554,34 +556,159 @@ static const ut8 *parse_line_header_source(RBinFile *bf, const ut8 *buf, const u
...
+					dwarf_read_offset (hdr->is_64bit, &buf, buf_end);
+					break;
+				case DW_FORM_strp:
+				case DW_FORM_line_strp:
+					section_name = form_code == DW_FORM_strp? "debug_str": "debug_line_str";
+					size_t section_len = 0;
+					ut64 section_offset = dwarf_read_offset (hdr->is_64bit, &buf, buf_end);
+					ut8 *section = get_section_bytes (bf->rbin, section_name, &section_len);
+					if (!section) {
+						// TODO handle this somehow
```
Comment (trufae) [inline libr/bin/dwarf.c:633]: handle this case to avoid a crash.

File: libr/bin/dwarf.c [hunk 6]
```diff
@@ -554,34 +556,159 @@ static const ut8 *parse_line_header_source(RBinFile *bf, const ut8 *buf, const u
...
+				case DW_LNCT_MD5:
+					// TODO Save the hash of the file.
+					break;
+				}
+			}
+
+			if (mode == R_MODE_PRINT && i == DIRECTORIES) {
+				// Keep directories 0 indexed?
+				print ("  %d     %s\n", index, sdb_array_get (sdb, "includedirs", index, 0));
+			} else if (mode == R_MODE_PRINT && i == FILES) {
```
Comment (trufae) [inline libr/bin/dwarf.c:694]: both cases use if (mode == R_MODE_PRINT. just group that like this:

```c
if (mode == R_MODE_PRINT) {
  switch (i) {
 case DIRECTORIES: 
  break;
case FILES:
 break;
}
```

**Review by trufae — CHANGES_REQUESTED** [review]
Will be good to do some fuzzing before merging this. I'm fine to merge it like this (but after 5.8.2) or do some fuzzing in `test/fuzz/` (see the readme to setup the thing). But the patch looks quite smaller than I expected. After addressing the comments i'll do another review. I can do some fuzzing on it if i find some time this week too. But having dwarf5 support was kind of high in the wishlist.

**Comment by MewtR** [comment]
I'll look into the fuzzing a little later this week.

**Comment by trufae** [comment]
Let's merge it! i will have some time today to fuzz it and test it :) so it will fit into the release! thank you!

**Pull request merged**